### DISCUSSION TITLE
How do I make Tailwind working?

### DISCUSSION THREAD

**Discussion by skylightsai** [imfing/hextra#111] [Q&A]
So I am overriding/changing styles of pages eg. `./layouts/blog/list.html` (my site's root, NOT inside `themes/hextra/layouts`) but the new Tailwind style classes are not being reflected on my site. 

To clarify: the existing Tailwind classes that comes with the theme are working out-of-the-box - but the classes I am adding are not.

How do I force Tailwind to refresh css so that it includes the styles of the classes? 

Looks like the command `hugo server` does not run alongside with Tailwind watch process? Am I missing something?

**Comment by imfing** [comment]
How Tailwind CSS classes work in Hextra is that - the classes are pre-compiled via PostCSS to [assets/css/compiled/main.css](https://github.com/imfing/hextra/blob/main/assets/css/compiled/main.css). When using the theme, the compilation process wouldn't kick in. 

In your case, if you are running `Tailwind watch process`, i.e. the `npx tailwindcss -i ./src/input.css -o ./dist/output.css --watch`, what you can do is setting the output directory under your `static` folder.
Then, you could include the compiled CSS file in [Custom Scripts | Hextra](https://imfing.github.io/hextra/docs/advanced/customization/#custom-scripts). Let me know if you have any questions.

**Reply by skylightsai** [reply]
Right. So:

1. Should I duplicate the `tailwind.config.js` and put it under my `root` dir too?
2. Do I need to duplicate entire `/assets/css/` directory and put it under `root` in order to have the `npx tailwind` process ALL necessary assets to run the theme (+ whatever new Tailwind classes or custom css I add)?
3. Given I will be modifying the original theme, wouldn't make more sense to simply make the theme my root instead of including it as as mod/git submodule?

**Reply by imfing** [reply]
Just for clarification, from my understanding, what you were trying to achieve was
1. create your own tailwind style file, e.g. `/assets/your-style.css`
2. compile the above stylesheet to e.g. `/static/compiled-style.css`, which will be loaded in your site to add or override some styles provided by Hextra

If that's the case then:

1. Yes, basically you could follow how I set up `tailwind.config.js` and `postcss.config.js` for Hextra: https://github.com/imfing/hextra
2.. I don't think you need to duplicate the entire `/assets/css/` as what Hextra needed has already been pre-compiled to [assets/css/compiled/main.css](https://github.com/imfing/hextra/blob/main/assets/css/compiled/main.css). You only need to use the `npx tailwind` to process your own classes added. for example, set the `content` to your own layouts:

```js
module.exports = {
  /* Content sources to tell Tailwind which files to scan for used class names for
     generating the needed CSS. More info:
     - https://tailwindcss.com/docs/content-configuration#configuring-source-paths
  */
  content: [
    "./layouts/**/*.html",
    "./content/**/*.{html,md}",
  ],
}
```
[source](https://foundata.com/en/blog/2023/hugo-theme-tailwind-css-jit-livereload/)

3. I don't recommend that as Hugo `layouts`/`assets`/`static` can be overridden by simply putting the same files under your site root. It would be difficult to upgrade theme in the future if you don't manage the theme using `git module` or `hugo module`

**Reply by Nu7ron** [reply]
Hi Xin

I was following the above instructions for compiling a custom.css using Tailwind. Unfortunately some of the tailwind classes of the hextra theme get overwritten, although I configured a very specific source folder to be considered only. This results in a broken navigation for example.

`
module.exports = {
  content: [
  "./layouts/shortcodes/*.html"
  ],
};
`

Also, I'm outputting the custom.css file to `/assets/css/custom.css` instead of `/static/custom.css` as stated in the your guide here [Custom Scripts](https://imfing.github.io/hextra/docs/advanced/customization/#custom-scripts).

I was wondering if you might be able to provide a little more details on how to create custom shortcodes using Tailwind? Do I need to use the full tailwind config from hextra in my config as well? Do I need postcss?

Currently my tailwind config is as minimal as possible to avoid dependencies to the config in the hextra theme.

**Reply by imfing** [reply]
> Hi Xin
> 
> 
> 
> I was following the above instructions for compiling a custom.css using Tailwind. Unfortunately some of the tailwind classes of the hextra theme get overwritten, although I configured a very specific source folder to be considered only. This results in a broken navigation for example.
> 
> 
> 
> `
> 
> module.exports = {
> 
>   content: [
> 
>   "./layouts/shortcodes/*.html"
> 
>   ],
> 
> };
> 
> `
> 
> 
> 
> Also, I'm outputting the custom.css file to `/assets/css/custom.css` instead of `/static/custom.css` as stated in the your guide here [Custom Scripts](https://imfing.github.io/hextra/docs/advanced/customization/#custom-scripts).
> 
> 
> 
> I was wondering if you might be able to provide a little more details on how to create custom shortcodes using Tailwind? Do I need to use the full tailwind config from hextra in my config as well? Do I need postcss?
> 
> 
> 
> Currently my tailwind config is as minimal as possible to avoid dependencies to the config in the hextra theme.

I think it could be some default tailwind css styles override Hextra's css style. You may need to use the tailwind config from this repo and modify the `content` path accordingly to generate your custom CSS file.

**Reply by Nu7ron** [reply]
Thanks for clarifying. Unfortunately I was not able to fix the overwrites with the full config.

What did fix the problem though is moving the `custom.css` include in `layout/partials/head-css.html` before the hextra css includes. This way all duplicates from custom.css get correctly overwritten by the main.css from the hextra theme.

I'm aware that this will not allow intended overwrites of tailwind classes in the hextra `main.css` anymore using `custom.css`, but do you think my changes in head-css.html would be beneficial to justify a commit?

**Reply by imfing** [reply]
> What did fix the problem though is moving the `custom.css` include in `layout/partials/head-css.html` before the hextra css includes. This way all duplicates from custom.css get correctly overwritten by the main.css from the hextra theme.
> 

The intended usage of `custom.css` is being able to overwrite the theme's built-in style.

> 
> I'm aware that this will not allow intended overwrites of tailwind classes in the hextra `main.css` anymore using `custom.css`, but do you think my changes in head-css.html would be beneficial to justify a commit?

I think swapping the order would cause problems for other users who don't use Tailwind CSS. 
I'll try figuring out a proper way to allow custom Tailwind classes.

**Reply by sakonn** [reply]
Hello, do you have any new input on this topic? 
I am trying to style a cookie banner for our website because of google analytics and I am struggling with custom Tailwind CSS too.

**Reply by imfing** [reply]
Hi @sakonn what issue did you encounter when using your own custom Tailwind CSS classes?

**Reply by sakonn** [reply]
Hello, I was trying to implement a custom partial for Google Analytics because we must have user consent via a cookie banner before analytics is loaded into the browser. So I am trying to style the banner with classes which are not in the main.css bundle out of the box. And some classes are not bundled there even when used in an HTML template. 

I have configured Hugo stats in my configuration:
```yaml
build:
  buildStats:
    enable: true
    disableIDs: true
  cachebusters:
    - source: assets/watching/hugo_stats\.json
      target: styles\.css
    - source: (postcss|tailwind)\.config\.js
      target: css
    - source: assets/.*\.(js|ts|jsx|tsx)
      target: js
    - source: assets/.*\.(.*)$
      target: $1

module:
  mounts:
    - source: assets
      target: assets
    - source: hugo_stats.json
      target: assets/watching/hugo_stats.json
```

I have quite simple banner template:
```html
<div class="border border-gray-200 bg-white text-gray-700
    dark:border-neutral-800 dark:bg-neutral-900 dark:text-gray-200
    md:rounded-lg z-10 fixed min-h-36 bottom-0 left-0 md:bottom-3 md:left-3 p-4 w-full md:w-96
    hidden"
    id="cookie-notice">
  <div class="leading-7 mt-2 mb-2">
    {{ i18n "cookieDesc" }}
  </div>
  <div class="mt-5 flex flex-row flex-wrap justify-end items-end">
    <button class="py-3 px-4 relative -ml-2 whitespace-nowrap p-2 font-medium" type="button" id="cookie-notice-deny">{{ i18n "cookieDecline" }}</button>
    <button class="py-3 px-4 relative -ml-2 whitespace-nowrap p-2 hover:text-gray-800 dark:hover:text-gray-200 btn-pri nav-btn" type="button" id="cookie-notice-accept">{{ i18n "cookieAccept" }}</button>
  </div>
</div>
```

The generated `hugo_stats.json` does contain all the classes but the `main.css` output does not contain them so the styling is not working as expected. For example `md:w-96`, `min-h-36`, `md:rounded-lg` classes are not present.

[main.css](https://github.com/imfing/hextra/files/14161935/main.txt)
[hugo_stats.json](https://github.com/imfing/hextra/files/14161936/hugo_stats.json)

**Reply by jgillich** [reply]
Also ran into the issue that Tailwind overrides theme styles. But what works for me is to simply use the JS
```
<script src="https://cdn.tailwindcss.com"></script>
```
This isn't deal for production. but better than nothing :smile: 

Edit: Just noticed this breaks styling in light mode. Too bad

Out of the box tailwind support would be sweet, I kinda expected that because the theme advertises using Tailwind

**Reply by jeddchoi** [reply]
@imfing I also would be grateful if you offer some guide to use tailwind css without modifying theme files.

**Reply by imfing** [reply]
Thank you all for the feedback @jgillich @jeddchoi @sakonn 

I have a [PR](https://github.com/imfing/hextra/pull/300) to add prefix to all Hextra's Tailwind utility classes, making it more seamless to use standard Tailwind CSS in your own project.

For example, out-of-the-box components can be copied as-is:

![image](https://github.com/imfing/hextra/assets/5097752/a5a35e66-64c0-4d27-afea-3de22b75fe3c)

it would give us:

![image](https://github.com/imfing/hextra/assets/5097752/ef305a50-fa87-41dd-adc9-e54b3306b854)

this is gonna unlock a lot of potential. I will figure out an elegant way to enable the production build from user side.

**Reply by imfing** [reply]
The latest main branch should have this change now, you could test it out by using the latest version:

```
hugo mod get -u github.com/imfing/hextra@main
```

**Reply by sakonn** [reply]
Thanks for the implementation. I updated the template to the latest version, and custom styling works fine now. Thank you very much. 🔝 🔝

**Comment by jeddchoi** [comment]
@imfing Hi, I found the official reference and sample project by creator of Hugo about using Tailwind CSS and Hugo.
I hope it would be helpful.

https://gohugo.io/getting-started/configuration/#configure-cache-busters
https://github.com/bep/hugo-starter-tailwind-basic

**Reply by imfing** [reply]
thanks, that's how this project has been set up ;)
